tools/tiller-hack/tiller-hack.go

```diff
@@ -0,0 +1,256 @@
+	}
+	if err != nil {
+		return nil, err
+	}
+	k8s, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return k8s, nil
+}
+
+func getConfigmapData(k8sClient *kubernetes.Clientset, configMapName, configMapNamespace string) ([]byte, error) {
+	cfgMap, err := k8sClient.Core().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("fail to get configmap %s/%s with error: %+v", configMapNamespace, configMapName, err)
+	}
+	data, ok := cfgMap.Data["release"]
+	if !ok {
+		return nil, fmt.Errorf("fail to locate \"release\" key in configmap %s/%s", configMapNamespace, configMapName)
+	}
+	return []byte(data), nil
+}
+
+func makeConfigMapBackup(k8sClient *kubernetes.Clientset, configMapName, configMapNamespace string) error {
+	cfgMap, err := k8sClient.Core().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("fail to get configmap %s/%s with error: %+v", configMapNamespace, configMapName, err)
+	}
+	bckpCfgMap := v1.ConfigMap{}
+	cfgMap.DeepCopyInto(&bckpCfgMap)
+	bckpCfgMap.ResourceVersion = ""
+
+	timestamp := strings.Replace(time.Now().Format("2006-01-0215:04:05.999"), "-", "", -1)
+	timestamp = strings.Replace(timestamp, ":", "", -1)
+	bckpCfgMap.ObjectMeta.Name += "." + timestamp + ".backup"
+	_, err = k8sClient.Core().ConfigMaps(configMapNamespace).Create(&bckpCfgMap)
+
+	return err
+}
+
+func decodeRelease(data []byte) (*rspb.Release, error) {
+	// base64 decode string
+	b, err := base64.StdEncoding.DecodeString(string(data))
+	if err != nil {
+		return nil, err
+	}
+
+	if bytes.Equal(b[0:3], magicGzip) {
+		r, err := gzip.NewReader(bytes.NewReader(b))
+		if err != nil {
+			return nil, err
+		}
+		b2, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+		b = b2
+	}
+
+	var rls rspb.Release
+	// unmarshal protobuf bytes
+	if err := proto.Unmarshal(b, &rls); err != nil {
+		return nil, err
+	}
+	return &rls, nil
+}
+
+func encodeRelease(rls *rspb.Release) (string, error) {
+	b, err := proto.Marshal(rls)
+	if err != nil {
+		return "", err
+	}
+	var buf bytes.Buffer
+	w, err := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	if err != nil {
+		return "", err
+	}
+	if _, err = w.Write(b); err != nil {
+		return "", err
+	}
+	w.Close()
+
+	return base64.StdEncoding.EncodeToString(buf.Bytes()), nil
+}
+
+func updateConfigMap(k8sClient *kubernetes.Clientset, configMapName, configMapNamespace, updatedData string) error {
+	cfgMap, err := k8sClient.Core().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("fail to get configmap %s/%s with error: %+v", configMapNamespace, configMapName, err)
+	}
+	cfgMap.Data["release"] = updatedData
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+		_, err = k8sClient.CoreV1().ConfigMaps(configMapNamespace).Update(cfgMap)
+		return
```

Comment:
how is this the user of this process notified of this failure event (or does that not matter, not familiar with RetryOnConflict).

Reply:
read the doc, very useful func for patching kubernetes api objects. Used a lot in k8s code.